File: leetcode-python-go/809. 情感丰富的文字/main.py
from typing import List

S = "heeellooo"
words = ["hello", "hi", "helo"]

def expressiveWords(s: str, words: List[str]) -> int:
    def expand(s: str, t: str) -> bool:
        i = j = 0
        while i < len(s) and j < len(t):
            if s[i] != t[j]:
                return False
            ch = s[i]
            cnti = 0
            while i < len(s) and s[i] == ch:
                cnti += 1
                i += 1
            cntj = 0
            while j < len(t) and t[j] == ch:
                cntj += 1
                j += 1
            
            if cnti < cntj:
                return False
            if cnti != cntj and cnti < 3:
                return False
        
        return i == len(s) and j == len(t)
    
    return sum(int(expand(s, word)) for word in words)



print(expressiveWords(S, words))

File: leetcode-python-go/809. 情感丰富的文字/main.go
package main

import (
	"fmt"
)

func expand(s, t string) bool {
	n, m := len(s), len(t)
	i, j := 0, 0
	for i < n && j < m {
		if s[i] != t[j] {
			return false
		}
		ch := s[i]
		cntI := 0
		for i < n && s[i] == ch {
			cntI++
			i++
		}
		cntJ := 0
		for j < m && t[j] == ch {
			cntJ++
			j++
		}
		if cntI < cntJ || cntI > cntJ && cntI < 3 {
			return false
		}
	}
	return i == n && j == m
}

func expressiveWords(s string, words []string) (ans int) {
	for _, word := range words {
		if expand(s, word) {
			ans++
		}
	}
	return
}

func main() {
	S := "heeellooo"
	words := []string{"hello", "hi", "helo"}
	results := expressiveWords(S, words)
	fmt.Println(results)
}
